### PR TITLE
+str Add mergeAll stream operator.

### PR DESCRIPTION
**Pull request by He-Pin**
References https://github.com/akka/akka/issues/19045

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala [hunk 1]
```diff
@@ -2890,6 +2890,30 @@ trait FlowOps[+Out, +Mat] {
...
+   *
+   * '''Emits when''' one of the inputs has an element available
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' all upstreams complete (eagerComplete=false) or one upstream completes (eagerComplete=true), default value is `false`
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def mergeAll[U >: Out](those: Graph[SourceShape[U], _]*)(eagerComplete: Boolean): Repr[U] = those match {
```
Comment (hepin1989) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:2905]: should this signature be curried or not @viktorklang @johanandren , or it should be ?
```scala
def mergeAll[U >: Out](eagerComplete: Boolean, those: Graph[SourceShape[U], _]*)
```

Comment (viktorklang) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:2905]: @hepin1989 I defer to @johanandren :)

Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:2905]: For the record, discussions going on in https://github.com/akka/akka/pull/31299#discussion_r840351828

Comment (hepin1989) [reply akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:2905]: I will reopen them later on

**Comment by hepin1989** [comment]
@raboof @johanandren I think this one is ready to review.

**Review by johanandren — COMMENTED** [review]
Looks good

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowMergeAllSpec.scala [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    "works in merge numbers example" in {
+      // #merge-all
+      val sourceA = Source(1 to 3)
+      val sourceB = Source(4 to 6)
+      val sourceC = Source(7 to 10)
+      sourceA.mergeAll(List(sourceB, sourceC), eagerComplete = false).runForeach(println)
+      // merging is not deterministic, can for example print 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
+      // #merge-all
+    }
+
```
[the remark this answers is not in the document]
Comment (He-Pin) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowMergeAllSpec.scala:57]: Seems reasonable, but that should apply to `concat` etc operators too.

[the remark this answers is not in the document]
Comment (johanandren) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowMergeAllSpec.scala:57]: I thought we did have that, but when looking closer the only static "combiner" source factory is `Source.combine`, which takes a strategy and already covers merging I guess.

**Review by johanandren — APPROVED** [review]

**Comment by He-Pin** [comment]
I will update other PRs once this gets merged.

**Comment by He-Pin** [comment]
@patriknw @raboof ping~~

**Comment by He-Pin** [comment]
have been rebased on to current master.

**Review by johanandren — APPROVED** [review]

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala [hunk 3]
```diff
@@ -1666,6 +1666,23 @@ final class Source[Out, Mat](delegate: scaladsl.Source[Out, Mat]) extends Graph[
...
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' all upstreams complete (eagerComplete=false) or one upstream completes (eagerComplete=true), default value is `false`
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def mergeAll(those: java.util.List[Source[Out, _ <: Any]], eagerComplete: Boolean): javadsl.Source[Out, Mat] = {
+    val seq = if (those != null) Util.immutableSeq(those).map(_.asScala) else immutable.Seq()
+    new Source(delegate.mergeAll(seq, eagerComplete))
```
Comment (He-Pin) [inline akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:1683]: Updated to javadsl's Source

Comment (He-Pin) [reply akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:1683]: or
```those: java.util.List[_<:Graph[SourceShape[Out], _<:Any]]```?

File: akka-stream-tests/src/test/java/akka/stream/javadsl/SourceTest.java [hunk 4]
```diff
@@ -1099,6 +1102,20 @@ public void apply(String elem) {
...
+    final Source<Integer, NotUsed> sourceA = Source.from(Arrays.asList(1, 2, 3));
+    final Source<Integer, NotUsed> sourceB = Source.from(Arrays.asList(4, 5, 6));
+    final Source<Integer, NotUsed> sourceC = Source.from(Arrays.asList(7, 8, 9));
+    final TestSubscriber.Probe<Integer> sub =
+        sourceA
+            .mergeAll(Arrays.asList(sourceB, sourceC), false)
+            .runWith(TestSink.probe(system), system);
+    sub.expectSubscription().request(9);
+    sub.expectNextUnorderedN(Util.immutableSeq(Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9)))
+        .expectComplete();
```
Comment (He-Pin) [inline akka-stream-tests/src/test/java/akka/stream/javadsl/SourceTest.java:1116]: I added a Java test

**Comment by He-Pin** [comment]
@patriknw I think this one is ready for review.

**Review by patriknw — APPROVED** [review]
LGTM

**Pull request merged**